Track RequestedRangeNotSatisfiedException separately in S3 Metrics

arteam opened the pull request:
Due to RCO changes, we started getting a lot of `RequestedRangeNotSatisfiedException`s which are expected. We would like  track them separately.

This change a new metrics `es.repositories.exceptions.request_range_not_satisfied.total`  to track all client errors analogous to other S3 errors.

In the future, we can add the error code as an attribute to the metrics, so we can adapt it to all client errors.

arteam committed:
Track RequestedRangeNotSatisfiedException separately in S3 Metrics

Due to RCO changes, we started getting a lot of `RequestedRangeNotSatisfiedException`s which
are expected. We would like to not count them as failures and track them separately.

arteam committed:
Update docs/changelog/109657.yaml

elasticsearchmachine commented:
Hi @arteam, I've created a changelog YAML for you.

arteam committed:
Don't create blobRepository and plugin for every test

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

arteam commented on modules/repository-s3/src/internalClusterTest/java/org/elasticsearch/repositories/s3/S3BlobStoreRepositoryMetricsTests.java:
```diff
@@ -166,6 +184,43 @@ public void testMetricsWithErrors() throws IOException {
+
+        for (int i = 0; i < randomIntBetween(1, 3); i++) {
+            final long batch = i + 1;
+            addErrorStatus(TOO_MANY_REQUESTS, TOO_MANY_REQUESTS, REQUESTED_RANGE_NOT_SATISFIED);
```
Ideally it would be great to test multiple `REQUESTED_RANGE_NOT_SATISFIED` errors in one request, but I'm not sure how it's possible.

ywangd commented on server/src/main/java/org/elasticsearch/repositories/RepositoriesMetrics.java:
```diff
@@ -28,10 +30,14 @@ public record RepositoriesMetrics(
 
     public static final String METRIC_REQUESTS_TOTAL = "es.repositories.requests.total";
     public static final String METRIC_EXCEPTIONS_TOTAL = "es.repositories.exceptions.total";
+    public static final String METRIC_EXCEPTIONS_REQUEST_RANGE_NOT_SATISFIED_TOTAL =
+        "es.repositories.exceptions.request_range_not_satisfied.total";
```
@idegtiarenko and I had some previous discussion about differentiate between server errors (e.g. 500) and client errors (e.g. 416 here and other 4xx errors like 404, 401, 403 etc). I wonder whether we should have the metrics to be named more generically for all client side errors, e.g. `es.repositories.exceptions.client.total`, and use attribute for the actual error code/name. It feels more expandable to me. What do you think?

arteam replied:
That's a great idea! I've updated the counter and the histogram to hold all client errors. Currently, I specifically only specifically discard 416 errors, but we can adjust the code in the future to handle all client errors.

ywangd reviewed (commented):
Thanks for working on this. I left a question.

arteam committed:
Merge remote-tracking branch 'origin/main' into track-requested-range-not-satisfied-exceptions

arteam committed:
Track all client errors separately from exceptions

arteam committed:
Upgrade names of the client error metrics

kingherc commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:
```diff
@@ -177,6 +179,30 @@ public final void collectMetrics(Request<?> request, Response<?> response) {
+                    .filter(e -> (Integer) e == REQUESTED_RANGE_NOT_SATISFIED.getStatus())
+                    .count();
+                if (amountOfRequestRangeNotSatisfiedErrors > 0) {
+                    // REQUESTED_RANGE_NOT_SATISFIED errors are expected errors due to RCO, so we would like not to count them along with
```
According to [this comment](https://elasticco.atlassian.net/browse/ES-8636?focusedCommentId=139000) by @henningandersen , it may be preferable to count them among the rest of the client errors, but just filter them out later (e.g., in Kibana's Discover or dashboards). So I wonder if it's better to not decrease the exceptionCount, but still pass on the clientErrorAttributes in the incrementBy? Will this allow us to filter out 416s?

ywangd replied:
IIUC, you are suggesting that we still maintain a single `es.repositories.exceptions.total` metric but with added attributes, e.g. error code, to faciliate differentiating them in reporting time. I think this could be a viable idea. It also defers the definition of client vs server errors at reporting time which seems like a good thing, i.e. extra flexibility. It feels one step further compared to [my earlier suggestion](https://github.com/elastic/elasticsearch/pull/109657#discussion_r1642111137).

In theory, each exception could have a different error status code. So we will need to record them separately here with this suggestion. But it seems worthwhile.

arteam replied:
I've adjusted the code to not to deduct the client error codes from the exception count. We just track client error codes separately with the error code as an attribute. I'm currently only tracking REQUESTED_RANGE_NOT_SATISFIED(416) errors, but happy to adjust to other client errors. 

Not sure if can track all errors, because errors like `TOO_MANY_REQUESTS` are also tracked in `statusCodes`.

kingherc commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:
```diff
@@ -177,6 +179,30 @@ public final void collectMetrics(Request<?> request, Response<?> response) {
                 .orElse(0);
 
+            if (exceptionCount > 0) {
+                final List<Object> statusCodes = awsRequestMetrics.getProperty(AWSRequestMetrics.Field.StatusCode);
```
Does this include only client status codes (4xx)?

arteam replied:
I believe it includes all status codes. For example, here's a trace from the `testMetricsWithErrors` test:

```
[2024-06-18T01:18:24,957][INFO ][o.e.r.s.S3BlobStore      ] [testMetricsWithErrors] Error codes: [429, 429, 200]
[2024-06-18T01:18:25,442][INFO ][o.e.r.s.S3BlobStore      ] [testMetricsWithErrors] Error codes: [429, 404]
[2024-06-18T01:18:25,778][INFO ][o.e.r.s.S3BlobStore      ] [testMetricsWithErrors] Error codes: [429, 404]
[2024-06-18T01:18:26,101][INFO ][o.e.r.s.S3BlobStore      ] [testMetricsWithErrors] Error codes: [429, 404]
[2024-06-18T01:18:31,672][INFO ][o.e.r.s.S3BlobStore      ] [testMetricsWithErrors] Error codes: [429, 429, 429, 429, 429]
```

kingherc replied:
200 (successful) is not an error though, so I hope we don't count it in the error metrics?

arteam replied:
We don't. Currently, we count only 416 errors from status codes.

kingherc reviewed (commented):
Nice. Some questions/comments.

arteam committed:
Update server/src/main/java/org/elasticsearch/repositories/RepositoriesMetrics.java

Co-authored-by: Iraklis Psaroudakis <kingherc@gmail.com>

arteam committed:
Run spotless

arteam committed:
Do not remove client errors from the total of exceptions

arteam committed:
Merge remote-tracking branch 'origin/main' into track-requested-range-not-satisfied-exceptions

kingherc commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:
```diff
@@ -177,6 +179,31 @@ public final void collectMetrics(Request<?> request, Response<?> response) {
                 .map(List::size)
                 .orElse(0);
 
+            if (exceptionCount > 0) {
```
I'm continuing the conversation on how the stats should look like here @ywangd @arteam . I believe an ideal approach would be to:

- `METRIC_EXCEPTIONS_TOTAL` counts the server-related errors (like 5xx)
- This PR's new `METRIC_CLIENT_EXCEPTIONS_TOTAL` to count the client-related errors (like 4xx)
- And to be able to somehow later filter out specific error codes (like 416) from the client-related errors

I'm not sure what's the best way to achieve that. I can think of a couple:

- Either a separate metric/LongCounter for each status error code.
- Or a "total" counter (as is now) but with added attribute(s) that should hopefully allows us to filter out specific status codes.

To avoid adding a gazillion counters for each status code, I have a preference for the latter approach. But unsure whether it can work in practice, i.e., when you go to Kibana later, that you can indeed easily e.g. filter out the 416s.

Thus, I'd suggest:

- Count all the client-related exceptions in `METRIC_CLIENT_EXCEPTIONS_TOTAL`.
- Add each client error code of `statusCodes` to the attributes, with their sum. E.g., if statusCodes is [429, 416, 429, 429], then the added attributes would be `key=429, value=3` and `key=416, value=1`.
- Spin up a custom cloud or serverless deployment out of this PR, and execute a workload that will provoke some of these. I believe a rally benchmark should be able to provoke some 416s.
- Go to Kibana and make two dashboard visualizations: (a) one that shows ALL client errors, and (b) one that shows/plots/sums only 416s. This will confirm we can do the filtering using the attributes.
- If we confirm the approach is working, we can do the same to add the attributes also to `METRIC_CLIENT_EXCEPTIONS_HISTOGRAM`, and I could also see adding server-related status codes attributes to `METRIC_EXCEPTIONS_TOTAL`/`METRIC_EXCEPTIONS_HISTOGRAM`.

What do you think?

arteam replied:
@kingherc I believe the current setup should work without additional counters.

* I'd like to keep this PR focused on `RequestedRangeNotSatisfiedExceptions`, but I'm happy to adjust the code to handle all client errors after we verify that counting 416 errors works. 

* Filtering by attributes works. The error code is added to metrics under the `labels` prefix and the metric has a name `es.repositories.exceptions.client.total.labels.error_code` with the value 416. We just adjust the KQL query add an additional clause.

* We already add the labels to the client error histogram. If this would work for 416 errors, I'm happy to adjust it to cover all 4xx error codes.

kingherc replied:
@arteam I see, thanks for explaining. Makes sense to limit this PR to just RequestedRangeNotSatisfiedExceptions!

However, the current PR is a bit "mixed". For example some points of incoherence:

- `METRIC_CLIENT_EXCEPTIONS_TOTAL` only counts 416s.
- `METRIC_EXCEPTIONS_TOTAL` includes client exceptions.

To make it simpler and clearer, maybe we should simply rename `METRIC_CLIENT_EXCEPTIONS_TOTAL` to `METRIC_REQUESTED_RANGE_NOT_SATISFIED_EXCEPTIONS_TOTAL` and similarly for the histogram one. What do you think? And in the future, if we'd like to capture all kinds of client exceptions separately, we can issue a new ticket and start a new discussion.

arteam replied:
@kingherc Sounds good to me! That was the name of metrics in the first iteration of this pull request, I've reverted the commits to it, but kept adding the error code as a metric attribute, so we can adapt the metrics to all client errors in the future.

ywangd replied:
I am not sure that I follow the decision here. IIUC, we agreed that different exceptions can be filtered/differentiated at reporting time (in Kibana). In that case, shouldn't we use a single metric, i.e. the existing `METRIC_EXCEPTIONS_TOTAL` for the exceptions and just add labels for the `error_code`. I don't quite understand why we decided to have a separate `METRIC_CLIENT_EXCEPTIONS_TOTAL`?

arteam replied:
@ywangd I guess we we wanted to have a clear separation between exceptions and client errors based on status codes.
The biggest issue is that we can multiple status codes, so you would need to aggregate them before adding as labels, and I'm not sure how that would work with histograms, for example.

I think the current implementation solves the issue in the Jira ticket and also verifies that attributes work. We can probably iterate on a generic solutions for client errors in a follow up issue?

kingherc replied:
@ywangd , @arteam , I would also prefer to holistically handle the error codes so we can filter them separately, but judging from the long conversations we're having on how to make that happen (whether with labels or separate counters or some other way), I believe it may be another ticket/PR altogether. 

So I think indeed that this PR that introduces just a separate counter for the REQUESTED_RANGE_NOT_SATISFIED exceptions is sufficient to resolve ES-8636.

What do you think?

kingherc replied:
Potentially we can create a new ticket for all error codes, and when it's done, we can remove the specific metrics about REQUESTED_RANGE_NOT_SATISFIED exceptions introduced by this PR. ?

arteam replied:
I totally agree! I think that is the best way forward. I can create a ticket to track all client errors with labels, but keep a dedicated `REQUESTED_RANGE_NOT_SATISFIED` metric for the time being which solves the issue in the ticket.

ywangd replied:
I still think using labels/attributes for differentiating the different exception is better. That said, it is something that we can refine relatively easily later. So I am onboard with the current approach.

arteam committed:
Revert "Upgrade names of the client error metrics"

This reverts commit 5210a1a1

arteam committed:
Revert "Track all client errors separately from exceptions"

This reverts commit aee54096

arteam committed:
Merge remote-tracking branch 'origin/main' into track-requested-range-not-satisfied-exceptions

kingherc commented on modules/repository-s3/src/internalClusterTest/java/org/elasticsearch/repositories/s3/S3BlobStoreRepositoryMetricsTests.java:
```diff
@@ -179,6 +234,16 @@ private long getLongCounterValue(TestTelemetryPlugin plugin, String instrumentNa
             .orElse(0L);
     }
 
+    private long getLongCounterValue(TestTelemetryPlugin plugin, String instrumentName, int errorCode, Operation operation) {
```
nit I think these two new functions you're adding now with the additional errorCode argument are not used anymore? So you can remove them from the PR.

kingherc commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:
```diff
@@ -177,6 +179,31 @@ public final void collectMetrics(Request<?> request, Response<?> response) {
+                if (amountOfRequestRangeNotSatisfiedErrors > 0) {
+                    var clientErrorAttributes = Maps.copyMapWithAddedEntry(
+                        attributes,
+                        "error_code",
```
nit I'm not sure the additional attribute for the "error_code" is necessary. Since we count the REQUESTED_RANGE_NOT_SATISFIED separately, we know they are the 416 error so there's no reason to redundantly store the error_code as an additional attribute.

arteam committed:
Remove adding labels for error codes

arteam commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into track-requested-range-not-satisfied-exceptions

kingherc reviewed (approved):
LGTM. However, I'd suggest to get one more approval, ideally @ywangd 's, before merging.

arteam commented:
Thank you very much @kingherc for a through review! Will wait for @ywangd  to give the PR another look :+1:

ywangd commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:
```diff
@@ -160,7 +161,7 @@ public final void collectMetrics(Request<?> request, Response<?> response) {
             final AWSRequestMetrics awsRequestMetrics = request.getAWSRequestMetrics();
             final TimingInfo timingInfo = awsRequestMetrics.getTimingInfo();
             final long requestCount = getCountForMetric(timingInfo, AWSRequestMetrics.Field.RequestCount);
-            final long exceptionCount = getCountForMetric(timingInfo, AWSRequestMetrics.Field.Exception);
+            long exceptionCount = getCountForMetric(timingInfo, AWSRequestMetrics.Field.Exception);
```
I think we can revert this change since we no longer take rangeNotSatisfied error away from it.

ywangd commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:
```diff
@@ -177,6 +178,26 @@ public final void collectMetrics(Request<?> request, Response<?> response) {
+                    s3RepositoriesMetrics.common()
+                        .requestRangeNotSatisfiedExceptionCounter()
+                        .incrementBy(amountOfRequestRangeNotSatisfiedErrors, attributes);
+                    s3RepositoriesMetrics.common()
+                        .requestRangeNotSatisfiedExceptionHistogram()
+                        .record(amountOfRequestRangeNotSatisfiedErrors, attributes);
```
I think histogram for `RangeNotSatisfiedException` is not actually useful. The existing exception histogram was added to measure how many retries s3 client normally need to eventually succeeds. Previously we suspected that the default 3 retries is not enough and increased it to 9. The histogram helped us to be sure it was not the case since we hardly have more than 1 retry and we changed it back to 3 (https://github.com/elastic/elasticsearch-serverless/pull/1352).

`RangeNotSatisfiedException` is not retryable. I'd expect it to always be 1. The request cannot be eventually successful after seeing this error. So it's not very useful and I think we can drop it.

arteam replied:
That's a great point!

arteam replied:
I think indeed the histogram is not very useful, since I couldn't find a use case where could have multiple 416 errors for one S3 action.

ywangd reviewed (approved):
LGTM

I have some minor comments for your consideration. Thanks!

arteam committed:
Make exceptionCount final again

arteam committed:
Remove histogram for REQUEST_RANGE_NOT_SATISFIED errors

It's not useful since request range requests are not retryable and we should have mostly one error

arteam commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into track-requested-range-not-satisfied-exceptions

arteam commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into track-requested-range-not-satisfied-exceptions
